discovery/sync_manager.go

```diff
@@ -0,0 +1,640 @@
+
+		// If the next pending active syncer is no longer in our lookup
+		// map, then the corresponding peer has disconnected, so we'll
+		// skip them.
+		if _, ok := m.pendingActiveSyncers[s.cfg.peerPub]; !ok {
+			next = m.pendingActiveSyncerQueue.Front()
+			continue
+		}
+
+		return s
+	}
+
+	return nil
+}
+
+// rotateActiveSyncers rotates all currently eligible active syncers.
+func (m *SyncManager) rotateActiveSyncers() {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	// If we don't have any candidates, we can return early.
+	if len(m.inactiveSyncers) == 0 {
+		log.Debug("No candidates to rotate active syncers")
+		return
+	}
+
+	// Otherwise, we do so we should attempt to rotate as many eligible
+	// active syncers as possible.
+	numRotated := 0
+	blacklist := make(map[routing.Vertex]struct{})
+	for _, currentSyncer := range m.activeSyncers {
+		// The active syncer must be in a chansSynced state in order to
+		// process sync transitions.
+		if currentSyncer.syncState() != chansSynced {
+			continue
+		}
+
+		// If we have an active syncer available, we'll go ahead and
+		// choose a random passive syncer to swap with. It's possible
+		// for us to not find a valid candidate, so we'll exit this loop
+		// in that case.
+		candidate := m.chooseRandomSyncer(blacklist)
+		if candidate == nil {
+			break
+		}
+
+		// Otherwise, we have both a valid active syncer and candidate
+		// to swap with, so we'll blacklist them to ensure they're not
+		// chosen once again.
+		blacklist[currentSyncer.cfg.peerPub] = struct{}{}
+		blacklist[candidate.cfg.peerPub] = struct{}{}
+
+		// We can then transition each syncer to their respective new
+		// sync type.
+		err := m.transitionSyncer(currentSyncer, PassiveSync)
+		if err != nil {
+			log.Errorf("Unable to transition GossipSyncer(%x) to "+
+				"%v: %v", currentSyncer.cfg.peerPub,
+				PassiveSync, err)
+			continue
+		}
+		_ = m.transitionSyncer(candidate, ActiveSync)
+		numRotated++
+	}
+
+	if numRotated == 0 {
+		log.Debug("No valid candidates to rotate active syncers")
+	} else {
+		log.Debugf("Rotated %d active syncers", numRotated)
+	}
+}
+
+// transitionSyncer transitions a syncer to a new syncType.
+//
+// NOTE: Transitions to an ActiveSync type are optimistic since we don't know
+// exactly when the transition will happen due to the round robin, therefore the
+// error does not need to be checked.
+func (m *SyncManager) transitionSyncer(s *GossipSyncer, newSyncType SyncerType) error {
+	log.Debugf("Requesting transition to %v for GossipSyncer(%x)",
+		newSyncType, s.cfg.peerPub)
+
+	// If we're transitioning to an ActiveSync type, we'll short-circuit the
+	// process by queueing the syncer in our round-robin. Once it's this
+	// syncer's turn, the transition will happen. If this transition fails,
+	// then any new gossip syncers will be used as active syncers until
+	// reaching NumActiveSyncers.
+	if newSyncType == ActiveSync {
```

Comment:
Since this method doesn't actually transitions syncers to the active state, consider handling that transition by calling `enqueueActiveSyncer` directly.

Reply:
What's wrong with the current approach? Would rather have one call in which a syncer can perform any transition.

Reply:
What is wrong is that the `syncerHandler` is sending messages _to itself_ with the current approach, which complicates the logic.

Maybe I'm confused about the distinction between active and pending-active syncers, but could it be enough to just let `rotateActiveSyncerCandidates` shuffle the pending-active map without doing any transitioning?

Reply:
Even if we do that, we'll still need to transition it once the ticker fires, so it'll send itself a message anyway. I agree that the handler sending messages to itself can be somewhat weird, but it seems like the only alternative is to have a separate handler responsible for just the tickers.